CoreStoreTests/MigrationTests.swift

```diff
@@ -0,0 +1,235 @@
+                case .failure(let error):
+                    XCTFail("\(error)")
+                }
+                expectation.fulfill()
+            })
+        })
+        
+        self.waitAndCheckExpectations()
+    }
+    
+    func test_ThatCustomSchemaMappingProvider_CanCopyEntityWithCustomEntityMapping() {
+        class Foo: CoreStoreObject {
+            var name = Value.Required<String>("name", initial: "")
+        }
+        
+        // Todo: The way this handles different version locks is flaky… It fails face on the ground in debug, but seems
+        // todo: to work fine in production, yet it's not clear if it transforms everything as expected.
+        
+        let src: CoreStoreSchema = CoreStoreSchema(modelVersion: "1", entities: [Entity<Foo>("Foo")])
+        let dst: CoreStoreSchema = CoreStoreSchema(modelVersion: "2", entities: [Entity<Foo>("Foo")])
+        
+        XCTAssertEqual(dst.rawModel().entities.first!.versionHash, src.rawModel().entities.first!.versionHash)
+        
+        let migration: CustomSchemaMappingProvider = CustomSchemaMappingProvider(from: "1", to: "2", entityMappings: [
+            .copyEntity(sourceEntity: "Foo", destinationEntity: "Foo")
+        ])
+        
+        /// Create the source store and data set.
+        withExtendedLifetime(DataStack(src), { stack in
+            try! stack.addStorageAndWait(SQLiteStore())
+            try! stack.perform(synchronous: { $0.create(Into<Foo>()).name.value = "Willy" })
+        })
+        
+        let expectation: XCTestExpectation = self.expectation(description: "migration-did-complete")
+        
+        withExtendedLifetime(DataStack(src, dst, migrationChain: ["1", "2"]), { stack in
+            _ = stack.addStorage(SQLiteStore(fileURL: SQLiteStore.defaultFileURL, migrationMappingProviders: [migration]), completion: {
+                switch $0 {
+                case .success(_):
+                    XCTAssertEqual(stack.modelSchema.rawModel().entities.count, 1)
+                    XCTAssertEqual(try! stack.fetchCount(From<Foo>()), 1)
+                    try! stack.perform(synchronous: { $0.create(Into<Foo>()).name.value = "Bobby" })
+                case .failure(let error):
+                    XCTFail("\(error)")
+                }
+                expectation.fulfill()
+            })
+        })
+        
+        self.waitAndCheckExpectations()
+    }
+    
+    func test_ThatCustomSchemaMappingProvider_CanTransformEntityWithCustomEntityMapping() {
+        class Foo: CoreStoreObject {
+            var name = Value.Required<String>("name", initial: "")
+            var futile = Value.Required<String>("futile", initial: "")
+        }
+        
+        class Bar: CoreStoreObject {
+            var firstName = Value.Required<String>("firstName", initial: "", renamingIdentifier: "name")
+            var lastName = Value.Required<String>("lastName", initial: "", renamingIdentifier: "placeholder")
```

Comment:
This should never migrate successfully as there is no "placeholder" key in the source entity.